src/path.c

```diff
@@ -0,0 +1,937 @@
+/**
+ * @file path.c
+ * @author Michal Vasko <mvasko@cesnet.cz>
+ * @brief Path functions
+ *
+ * Copyright (c) 2020 CESNET, z.s.p.o.
+ *
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+#define _ISOC99_SOURCE /* strtoull */
+
+#include "path.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <stdlib.h>
+
+#include "common.h"
+#include "log.h"
+#include "plugins_types.h"
+#include "tree_data_internal.h"
+#include "tree_schema_internal.h"
+#include "xpath.h"
+
+/**
+ * @brief Check predicate syntax.
+ *
+ * @param[in] ctx libyang context.
+ * @param[in] exp Parsed predicate.
+ * @param[in,out] exp_idx Index in @p exp, is adjusted.
+ * @param[in] prefix Prefix option.
+ * @param[in] pred Predicate option.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+ly_path_check_predicate(const struct ly_ctx *ctx, const struct lyxp_expr *exp, uint16_t *exp_idx, uint8_t prefix,
+                        uint8_t pred)
+{
+    struct ly_set *set = NULL;
+    uint32_t i;
+    const char *name;
+    size_t name_len;
+
+    if (!lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_BRACK1)) {
+        /* '[' */
+        ++(*exp_idx);
+
+        if (((pred == LY_PATH_PRED_SIMPLE) || (pred == LY_PATH_PRED_KEYS))
+                && !lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_NAMETEST)) {
+            set = ly_set_new();
+            LY_CHECK_ERR_GOTO(!set, LOGMEM(ctx), error);
+
+            goto simple_nametest;
+            do {
+                /* ']' */
+                ++(*exp_idx);
```

Comment:
This is connected to the first conversation and it is this way because we do not know beforehand whether another predicate follows or not, for that both `]` and `[` tokens must be found, so they are simply skipped here.